### PR TITLE
Issue #3440: Added new check for verifying that annotation located on same line with its target

### PR DESCRIPTION
**Pull request by kazachka**
Issue #3440 

Added new check for verifying that annotation located on same line with its target.

### CODE REVIEW

File: src/main/resources/com/puppycrawl/tools/checkstyle/checks/annotation/messages.properties [hunk 1]
```diff
@@ -6,10 +6,10 @@ annotation.missing.override=Must include @java.lang.Override annotation when '{'
...
 annotation.parens.missing=Annotation must have closing parenthesis.
 annotation.parens.present=Annotation cannot have closing parenthesis.
+annotation.same.line=Annotation ''{0}'' should be on the same line with its target.
 annotation.trailing.comma.missing=Annotation array values must contain trailing comma.
 annotation.trailing.comma.present=Annotation array values cannot contain trailing comma.
 javadoc.duplicateTag=Duplicate {0} tag.
 javadoc.missing=Missing a Javadoc comment.
 suppressed.warning.not.allowed=The warning ''{0}'' cannot be suppressed at this location.
 tag.not.valid.on=The Javadoc {0} tag is not valid at this location.
-
```
Comment (kazachka) [inline src/main/resources/com/puppycrawl/tools/checkstyle/checks/annotation/messages.properties]: deleted
now there is different empty lines in different language files

File: src/xdocs/config_annotation.xml [hunk 2]
```diff
@@ -228,6 +228,111 @@ public void test(&#64;MyAnnotation String s) { // OK
       </subsection>
     </section>
 
+    <section name="AnnotationOnSameLine">
+      <subsection name="Description">
+        <p>Since Checkstyle 8.1</p>
```
Comment (kazachka) [inline src/xdocs/config_annotation.xml]: updated

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/annotation/annotationonsameline/InputAnnotationOnSameLineCheck.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+public class InputAnnotationOnSameLineCheck {
+
+    @Annotation int x;
+
+    int y;
+
+    @Annotation
+    @SomeClass.Annotation
+    public int getX() {
```
Comment (kazachka) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/annotation/annotationonsameline/InputAnnotationOnSameLineCheck.java]: added cases from issue and for all of tokens except of DOT (i can't find any usage of annotation with this token)

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/annotation/AnnotationOnSameLineCheckTest.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        };
+        assertArrayEquals("Default acceptable tokens are invalid", expected, actual);
+    }
+
+    @Test
+    public void testCheck() throws Exception {
+        final DefaultConfiguration config = createModuleConfig(AnnotationOnSameLineCheck.class);
+        final String[] expected = {
+            "9: " + getCheckMessage(MSG_KEY_ANNOTATION_ON_SAME_LINE, "Annotation"),
+            "10: " + getCheckMessage(MSG_KEY_ANNOTATION_ON_SAME_LINE, "SomeClass"),
```
Comment (kazachka) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/annotation/AnnotationOnSameLineCheckTest.java]: Method updated.
Note that i copy it from AnnotationLocationCheck, so AnnotationLocation can wrote "SomeClass" instead "Annotation".
Added java.lang.Deprecated as example.

File: src/xdocs/config_annotation.xml [hunk 5]
```diff
@@ -228,6 +228,111 @@ public void test(&#64;MyAnnotation String s) { // OK
...
+            </td>
+            <td>
+              <a href="apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#CLASS_DEF">CLASS_DEF</a>,
+              <a href="apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#INTERFACE_DEF">INTERFACE_DEF</a>,
+              <a href="apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#ENUM_DEF">ENUM_DEF</a>,
+              <a href="apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#METHOD_DEF">METHOD_DEF</a>,
+              <a href="apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#CTOR_DEF">CTOR_DEF</a>,
+              <a href="apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#VARIABLE_DEF">VARIABLE_DEF</a>.
+            </td>
+            <td>8.1</td>
```
Comment (kazachka) [inline src/xdocs/config_annotation.xml]: updated

**Comment by rnveach** [comment]
@kazachka Shippable failed.
> -Ppitest-checks-annotation
> [ERROR] Failed to execute goal org.pitest:pitest-maven:1.2.2:mutationCoverage (default-cli) on project checkstyle: Mutation score of 93 is below threshold of 100 -> [Help 1]

**Comment by romani** [comment]
@kazachka , please make sure that there is 100% coverage on your Check, in pitest report.

**Comment by kazachka** [comment]
I just forgot to add test to pitest configuration in pom.xml

File: pom.xml [hunk 6]
```diff
@@ -1675,6 +1675,7 @@
             <configuration>
               <targetClasses>
                 <param>com.puppycrawl.tools.checkstyle.checks.annotation.AnnotationLocationCheck</param>
```
Comment (rnveach) [inline pom.xml]: @romani Why wasn't this `com.puppycrawl.tools.checkstyle.checks.annotation.*`?

Comment (romani) [inline pom.xml]: because of https://github.com/hcoles/pitest/issues/368

**Comment by romani** [comment]
@kazachka , please provide testing report of this Check over multiple projects, please use our checkstyle-tester tool

File: src/xdocs/config_annotation.xml [hunk 7]
```diff
@@ -228,6 +228,111 @@ public void test(&#64;MyAnnotation String s) { // OK
       </subsection>
     </section>
 
+    <section name="AnnotationOnSameLine">
+      <subsection name="Description">
+        <p>Since Checkstyle 8.2</p>
+        <p>
+          Check annotation located on same line with its target.
```
Comment (romani) [inline src/xdocs/config_annotation.xml]: We need make a note that this Check is not a good practice, but some style guides use it.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/AnnotationOnSameLineCheck.java [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+        }
+    }
+
+    /**
+     * Finds next node of ast tree.
+     * @param node current node
+     * @return node that is next to given
+     */
+    private static DetailAST getNextNode(DetailAST node) {
+        final DetailAST nextNode;
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/AnnotationOnSameLineCheck.java]: All your other forms of writing this was writing the assignment with the standard one, and than change it's assignment if it was null. Why did you change it here?
````
DetailAST nextNode = node.getNextSibling();
if (nextNode == null) {
    nextNode = node.getParent().getNextSibling();
}
````

File: src/xdocs/checks.xml [hunk 9]
```diff
@@ -48,6 +48,10 @@
         <td>Check location of annotation on language elements.</td>
       </tr>
       <tr>
+        <td><a href="config_annotation.html#AnnotationOnSameLine">AnnotationOnSameLine</a></td>
+        <td>Check annotation on language elements is on same line with its target element.</td>
```
Comment (rnveach) [inline src/xdocs/checks.xml]: Why isn't this the same text as in the check in `config_annotation` or in the main Javadoc for the check?

File: src/xdocs/config_annotation.xml (hunk 7 above)
Comment (rnveach) [inline src/xdocs/config_annotation.xml]: Please make this description the same the check's javadoc.

Comment (rnveach) [reply src/xdocs/config_annotation.xml]: Make sure this note is in the check's Javadoc.

Comment (kazachka) [reply src/xdocs/config_annotation.xml]: Note added to both xdocs and javadoc

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/AnnotationOnSameLineCheck.java (hunk 8 above)
Comment (kazachka) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/AnnotationOnSameLineCheck.java]: changed

File: src/xdocs/checks.xml (hunk 9 above)
Comment (kazachka) [reply src/xdocs/checks.xml]: Updated to javadoc text

File: src/xdocs/config_annotation.xml (hunk 7 above)
Comment (kazachka) [reply src/xdocs/config_annotation.xml]: Updated to javadoc text

**Comment by kazachka** [comment]
Diff report:
https://kazachka.github.io/single-line-annotation-diff-report/

**Comment by romani** [comment]
As soon as TC passed , we could merge. but lets wait for TC to finish.
